Adding support for cassandra's CREATE ROLE

The-Alchemist opened the pull request:
documentation at https://cassandra.apache.org/doc/4.0/cql/security.html

Needs some more testing

### What does this PR do?
Adds support for cassandra's `CREATE ROLE`

### What issues does this PR fix or reference?

None
### Tests written?

No.  Not sure where the tests should go.  https://github.com/saltstack/salt/blob/develop/tests/unit/modules/test_cassandra_cql.py ?

### Commits signed with GPG?

No

The-Alchemist committed:
Adding support for cassandra's CREATE ROLE

documentation at https://cassandra.apache.org/doc/4.0/cql/security.html

Needs some more testing

cachedout commented:
@The-Alchemist Thanks for this! There are some lint warnings which need to be corrected before we can merge this, however. They can be reviewed here: https://jenkinsci.saltstack.com/job/pr-lint/job/PR-49829/2/warnings52Result/new/

Once you've pushed those changes, let us know and we'll take another look. Thanks!

gtmanfred commented on salt/modules/cassandra_cql.py:
```diff
@@ -882,6 +882,60 @@ def create_user(username, password, superuser=False, contact_points=None, port=N
+        cql_query(query, contact_points, port, cql_user, cql_pass)
+    except CommandExecutionError:
+        log.critical('Could not create role.')
+        raise
```
Can we just return False here instead of raising the exception?

That way people can handle the return if it was unable to create the role.

The-Alchemist replied:
thank you, @gtmanfred .  I wanted to be consistent with the rest of the code, and it seems every function raises some kind of exception on error, especially from `cql_query()`.

The-Alchemist replied:
@gtmanfred : just to clarify, you'd like this PR to change ALL the `cql` functions to just return False?

gtmanfred commented on salt/modules/cassandra_cql.py:
```diff
@@ -882,6 +882,60 @@ def create_user(username, password, superuser=False, contact_points=None, port=N
+        raise
+    except BaseException as e:
+        log.critical('Unexpected error while creating role: %s', e)
+        raise
```
same here.

The-Alchemist committed:
Merge branch 'develop' into patch-1

The-Alchemist committed:
Fixing formatting issues

From https://jenkinsci.saltstack.com/job/pr-lint/job/PR-49829/2/warnings52Result/new/

The-Alchemist committed:
Merge branch 'develop' into patch-1

gtmanfred commented:
In another PR that also adds handling that in the state, yes.

On Wed, Oct 10, 2018, 12:15 PM The Alchemist <notifications@github.com>
wrote:

> *@The-Alchemist* commented on this pull request.
> ------------------------------
>
> In salt/modules/cassandra_cql.py
> <https://github.com/saltstack/salt/pull/49829#discussion_r224166109>:
>
> > +        salt 'minion1' cassandra_cql.create_role username=joe password=secret superuser=True login=True
> +
> +        salt 'minion1' cassandra_cql.create_role username=joe password=secret superuser=True login=True contact_points=minion1
> +    '''
> +    superuser_cql = 'superuser = true' if superuser else 'superuser = false'
> +    login_cql     = 'login = true' if login else 'login = false'
> +    query = '''create role if not exists {0} with password '{1}' and {2} and {3} ;'''.format(username, password, superuser_cql, login_cql)
> +    log.debug("Attempting to create a new role with username=%s superuser=%s login=%s", username, superuser, login)
> +
> +    # The create role query doesn't actually return anything if the query succeeds.
> +    # If the query fails, catch the exception, log a messange and raise it again.
> +    try:
> +        cql_query(query, contact_points, port, cql_user, cql_pass)
> +    except CommandExecutionError:
> +        log.critical('Could not create role.')
> +        raise
>
> @gtmanfred <https://github.com/gtmanfred> : just to clarify, you'd like
> this PR to change ALL the cql functions to just return False?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/49829#discussion_r224166109>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAssoUo3xVUNIHucwbPFYL0o6vc8-7X4ks5ujitLgaJpZM4W_n6l>
> .
>

The-Alchemist commented:
> In another PR that also adds handling that in the state, yes.

@gtmanfred : I'm sorry, I'm not familiar with that part of the salt code.  What does "handling that in the state" mean?  Could you point me to an example?

Also, are there any other changes you'd like to see  in _this_ PR?

gtmanfred commented:
Looks like there is not a state module, so nevermind on that.

I would like to see the return changed to False here, instead of a stack trace, and then in another PR fix all the other states in this module, because we should be returning results that are possible to evaluate, instead of just exiting with an exception.

gtmanfred commented:
you could also return a dictionary or a message explaining why the failure happened, but raising an exception i don't think is appropriate.

cachedout commented:
@The-Alchemist Are you going to be able to come back to this one?

The-Alchemist commented:
@cachedout : Probably not this week, so if you wanna run with it, go for it. :)

cachedout commented:
@The-Alchemist That's fine. We'll check back with you next week. Thanks!

cachedout commented:
@The-Alchemist Just checking in with you again. :) Let us know if you're able to come back to this.

The-Alchemist commented:
@cachedout : sorry, I'm out of bandwidth!  It's all yours.

Commit:
Merge branch 'develop' into patch-1

thatch45 commented:
Thanks @The-Alchemist !
